// osu/dist/beatmap/structs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Mode;
(function (Mode) {
    Mode[Mode["Standard"] = 0] = "Standard";
    Mode[Mode["Taiko"] = 1] = "Taiko";
    Mode[Mode["Catch"] = 2] = "Catch";
    Mode[Mode["Mania"] = 3] = "Mania";
})(Mode = exports.Mode || (exports.Mode = {}));
var HitsoundType;
(function (HitsoundType) {
    HitsoundType[HitsoundType["None"] = 0] = "None";
    HitsoundType[HitsoundType["Normal"] = 1] = "Normal";
    HitsoundType[HitsoundType["Soft"] = 2] = "Soft";
    HitsoundType[HitsoundType["Drum"] = 3] = "Drum";
})(HitsoundType = exports.HitsoundType || (exports.HitsoundType = {}));
var AdditionType;
(function (AdditionType) {
    AdditionType[AdditionType["None"] = 0] = "None";
    AdditionType[AdditionType["Whistle"] = 1] = "Whistle";
    AdditionType[AdditionType["Finish"] = 2] = "Finish";
    AdditionType[AdditionType["Clap"] = 3] = "Clap";
})(AdditionType = exports.AdditionType || (exports.AdditionType = {}));
var CurveType;
(function (CurveType) {
    CurveType[CurveType["Unknown"] = 0] = "Unknown";
    CurveType[CurveType["Linear"] = 1] = "Linear";
    CurveType[CurveType["Catmull"] = 2] = "Catmull";
    CurveType[CurveType["Bezier"] = 3] = "Bezier";
    CurveType[CurveType["Perfect"] = 4] = "Perfect";
})(CurveType = exports.CurveType || (exports.CurveType = {}));
var Mods;
(function (Mods) {
    Mods[Mods["None"] = 0] = "None";
    Mods[Mods["NoFail"] = 1] = "NoFail";
    Mods[Mods["Easy"] = 2] = "Easy";
    Mods[Mods["NoVideo"] = 4] = "NoVideo";
    Mods[Mods["Hidden"] = 8] = "Hidden";
    Mods[Mods["HardRock"] = 16] = "HardRock";
    Mods[Mods["SuddenDeath"] = 32] = "SuddenDeath";
    Mods[Mods["DoubleTime"] = 64] = "DoubleTime";
    Mods[Mods["Relax"] = 128] = "Relax";
    Mods[Mods["HalfTime"] = 256] = "HalfTime";
    Mods[Mods["Nightcore"] = 512] = "Nightcore";
    Mods[Mods["Flashlight"] = 1024] = "Flashlight";
    Mods[Mods["Autoplay"] = 2048] = "Autoplay";
    Mods[Mods["SpunOut"] = 4096] = "SpunOut";
    Mods[Mods["Relax2"] = 8192] = "Relax2";
    Mods[Mods["Perfect"] = 16384] = "Perfect";
    Mods[Mods["Key4"] = 32768] = "Key4";
    Mods[Mods["Key5"] = 65536] = "Key5";
    Mods[Mods["Key6"] = 131072] = "Key6";
    Mods[Mods["Key7"] = 262144] = "Key7";
    Mods[Mods["Key8"] = 524288] = "Key8";
    Mods[Mods["keyMod"] = 1015808] = "keyMod";
    Mods[Mods["FadeIn"] = 1048576] = "FadeIn";
    Mods[Mods["Random"] = 2097152] = "Random";
    Mods[Mods["LastMod"] = 4194304] = "LastMod";
    Mods[Mods["FreeModAllowed"] = 2077883] = "FreeModAllowed";
    Mods[Mods["Key9"] = 16777216] = "Key9";
    Mods[Mods["Key10"] = 33554432] = "Key10";
    Mods[Mods["Key1"] = 67108864] = "Key1";
    Mods[Mods["Key3"] = 134217728] = "Key3";
    Mods[Mods["Key2"] = 268435456] = "Key2";
})(Mods = exports.Mods || (exports.Mods = {}));
